### DISCUSSION TITLE
Fundamentos do ASP.NET 6 > CRUD e ENTITY FRAMEWORK

### DISCUSSION THREAD

**Discussion by DrenokBR** [balta-io/forum#467] [Tira Dúvidas] [answered]
https://github.com/balta-io/2811/blob/main/Modulo%203/Blog/Controllers/CategoryController.cs

Na linha 24 do código disponível pela URL acima, referente ao Curso "Fundamentos do ASP.NET 6", é obrigatório definir a tipagem '<List<Category>>' na instanciação do objeto 'ResultViewModel' considerando que esteja sendo usada uma sobrecarga que tem apenas o tipo string como retorno ??? E se for obrigatório, qual o motivo ??? E seria possível definir vários tipos de retorno diferentes na mesma classe Controller para instâncias diferentes de 'ResultViewModel', já que é uma classe do tipo "Generic" ???

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @DrenokBR, como vai?

Não sei se entendi bem o que você quer fazer (Se puder explicar colocando trechos de código, com mais detalhes, fica melhor)...
Mas em resumo, se precisar retornar apenas uma string, não precisa do LIST, agora se for uma lista de strings aí precisa.

Sobre definir vários tipos de retorno, é possível sim, cria uma classe com estes tipos e informa o tipo dela no retorno, então ao invés de retornar uma lista ou uma string, você retorna uma classe com todos os dados.

Era isto?

**Comment by DrenokBR** [comment]
Postei a pergunta faltando uma parte importante... na segunda linha da minha pergunta ficou escrito 'List', mas na verdade era "ListCategory" assim como está na linha 24 do código do seu repositório do Github.

Mas a minha dúvida surgiu porque na linha 24 você chama a sobrecarga 'ResultViewModel(string error)' classe 'ResultViewModel' que não parece ter nenhuma relação com o tipo ''ListCategory' . Não consegui entender muito bem essa parte.

**Reply by DrenokBR** [reply]
Obs.: O github está cortando partes dos códigos mencionados... Decidi suprimir os caracteres < >

**Reply by DrenokBR** [reply]
Essa daqui tá valendo (rsrs...). Ainda fiquei na dúvida...

**Reply by andrebaltieri** [reply]
Bom dia, @DrenokBR, como vai?

Acredito que sua dúvida seja em relação aos Generics que explico no curso de Orientação a Objetos.

Se você inspecionar a classe ResultViewModel, verá na linha 3 que ela permite o recebimento de um tipo genérico `T`:
https://github.com/balta-io/2811/blob/main/Modulo%206/Blog/ViewModels/ResultViewModel.cs#L3

Isto torna possível passar qualquer tipo para esta classe!

Desta forma, se você passa uma `string` a propriedade `Data` do `ResultViewModel` será do tipo `string`, conforme linha 26:
https://github.com/balta-io/2811/blob/main/Modulo%206/Blog/ViewModels/ResultViewModel.cs#L26

Por isto esta mutação é possível:
`return StatusCode(500, new ResultViewModel<List<Category>>("05X04 - Falha interna no servidor"));`

**Reply by DrenokBR** [reply]
Ok. Obrigado, André.
